components/style/supports.rs

```diff
@@ -0,0 +1,211 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! [@supports rules](https://drafts.csswg.org/css-conditional-3/#at-supports)
+
+use cssparser::Parser;
+use parser::ParserContext;
+use properties::{PropertyDeclaration, PropertyId};
+use std::fmt;
+use style_traits::ToCss;
+
+#[derive(Debug)]
+/// An @supports condition
+///
+/// https://drafts.csswg.org/css-conditional-3/#at-supports
+pub enum SupportsCondition {
+    /// `not (condition)`
+    Not(Box<SupportsCondition>),
+    /// `(condition) and (condition) and (condition) ..`
+    And(Vec<SupportsCondition>),
+    /// `(condition) or (condition) or (condition) ..`
+    Or(Vec<SupportsCondition>),
+    /// `property-ident: value` (value can be any tokens)
+    Declaration(Declaration),
+    /// `(any tokens)` or `func(any tokens)`
+    FutureSyntax(String),
+}
+
+impl SupportsCondition {
+    /// Parse a condition
+    ///
+    /// `toplevel` is true when parsing a top-level condition in an `@supports` block,
+    /// and false when dealing with paren-nested conditions like `not(cond)` or `(cond) and (cond)`.
+    /// Declaration and FutureSyntax conditions can't appear in the top level without parens
+    pub fn parse(input: &mut Parser, toplevel: bool) -> Result<SupportsCondition, ()> {
+        if let Ok(_) = input.try(|i| i.expect_ident_matching("not")) {
+            let inner = SupportsCondition::parse_with_parens(input)?;
+            return Ok(SupportsCondition::Not(Box::new(inner)));
+        }
+        if !toplevel {
+            if let Ok(decl) = input.try(Declaration::parse) {
+                return Ok(SupportsCondition::Declaration(decl))
+            }
+        }
+
+        if let Ok(first_cond) = input.try(SupportsCondition::parse_with_parens) {
+            if input.is_exhausted() {
+                // nested parens
```

Comment:
Made it explicitly handle parenthesized cases, which fixes a couple tests.

@heycam suggested never storing it in a structured form in the first place, since we only need it for `ToCss`ing -- we can just evaluate whilst parsing. However, the spec suggests we also clean up the whitespace and such, and it's better to store it in the structured form for that.

(Also, I bet eventually there will be more structured CSSOM access to this like we have for media queries)